### PR TITLE
Add ability to pass down userPreferences to the LSP through the CLI

### PR DESCRIPTION
**Pull request by rmagatti**
Draft work for adding the ability to pass down `UserPreferences` to the LSP.

### CODE REVIEW

File: server/src/lsp-server.ts [hunk 1]
```diff
@@ -93,91 +93,94 @@ export class LspServer {
...
-            logger: this.options.logger,
-            onEvent: this.onTsEvent.bind(this)
-        });
-
-        this.tspClient.start();
-        this.tspClient.request(CommandTypes.Configure, {
-            preferences: {
-                allowTextChangesInNewFiles: true
+    initialize(preferences?: tsp.UserPreferences) {
+        return async function(params: TypeScriptInitializeParams): Promise<TypeScriptInitializeResult> {
```
Comment (rmagatti) [inline server/src/lsp-server.ts:97]: There's probably a better way to do this, but for now I just curried the initialize function to pass down the user preferences.

File: server/src/lsp-server.ts [hunk 2]
```diff
@@ -93,91 +93,94 @@ export class LspServer {
...
-                hoverProvider: true,
-                renameProvider: true,
-                referencesProvider: true,
-                signatureHelpProvider: {
-                    triggerCharacters: ['(', ',', '<']
+            this.tspClient.start();
+            this.tspClient.request(CommandTypes.Configure, {
+                preferences: {
+                    allowTextChangesInNewFiles: true,
+                    ...preferences
```
Comment (rmagatti) [inline server/src/lsp-server.ts:129]: Used here.

Every other change in this file is merely identation.

**Comment by rchl** [comment]
Why are those options not passed through the `initialize` request and instead on the command line? Is there something preventing it?

**Comment by rmagatti** [comment]
> Why are those options not passed through the `initialize` request and instead on the command line? Is there something preventing it?

Hm, probably just my lack of understanding?
First time even looking at this code really, just needed to make this work for my own local workflow, thought I'd put at least a draft up to get precisely this kind of feedback. I admittedly do not understand this codebase as much as I'd like to yet :)

I can dig into it a bit more and change the approach if something else makes more sense though 😊

**Comment by rmagatti** [comment]
#218 has a better implementation than this

**Pull request closed without merging**